### ISSUE TITLE
problems scaling fio --client and hence pbench-fio

### ISSUE THREAD

**Issue by bengland2** [distributed-system-analysis/pbench#360]
Sometimes, when I try to run fio --client on a set of 112 ext4 filesystems spread across 7 OSD hosts, each backed by a Ceph RBD device, I get [this error](http://pastebin.test.redhat.com/406992) .  The root of the problem appears to be that ETA messages from the workload generators are not reaching the test driver in time, they time out on the test driver, and subsequently the test driver sees some sort of field in the ETA message that it does not expect to see.  The job file is available [here](http://git.app.eng.bos.redhat.com/git/perf-dept.git/tree/scripts/ceph-ansible/pbench-fio-write.job) .

Other weird behaviors:  the estimated time remaining in the test increases from 5 min to 14 min and then very quickly back down to under a minute.  So obviously this is not working as expected.  The load was generated by:

    # ./fio.sh 16 fio y

on a set of filesystems generated with:

    # map-rbds.sh ../osds.list 16

Scripts available from [here](http://git.app.eng.bos.redhat.com/git/perf-dept.git/tree/scripts/ceph-ansible).

**Comment by portante** [comment]
@bengland2, I don't believe those links are reachable by folks external to Red Hat.  Can you edit to make that data public, properly redacting where necessary?  Thanks.

**Comment by bengland2** [comment]
sorry, [try here instead](http://pastebin.com/f0SbaqPd) for the fio output.  Looking at fio code in client.c, you can see that the assertion is preceded by a "timeout on SEND_ETA" message.  If you look at where that message is emitted by fio, you can see that it sets the client->eta_in_flight field to NULL.  My hypothesis is that this triggers the assertion, confirmed by gdb.    I'm going to discuss with the fio developers and see what they think.  IMHO this seems solvable because ETA messages are not required for fio workload generators to run the test -- they just report back on progress being made, and their frequency and timeout may be too aggressive, probably not taking into account the use of fio --client at scale as we intend.

**Comment by bengland2** [comment]
fix available [here](https://github.com/axboe/fio/commit/308d69b5d340577b7886696f39753b7ba5ae9e11) from Jens Axboe.

**Comment by ndokos** [comment]
Fix incorporated into pbench-fio 2.15-6 in COPR pbench-test repo.

**Comment by bengland2** [comment]
from pbench-fio-2.14-7.x86_64, when I run the jobfile below, I get:

    -=>>fio --client=c07-h01-6048r f2.job 
    hostname=c07-h01-6048r.rdu.openstack.engineering.redhat.com, be=0, 64-bit, os=Linux, arch=x86-64, fio=fio-2.14, flags=0
    <c07-h01-6048r.rdu.openstack.engineering.redhat.com> fio: --write_hist_log requires zlib in client/server mode
    client: host=c07-h01-6048r disconnected

    -=>>cat f2.job
    [global]
    write_bw_log=fio
    write_iops_log=fio
    write_lat_log=fio
    log_avg_msec=1000
    write_hist_log=fio
    log_hist_msec=1000
    log_hist_coarseness=4 # 76 bins
    [job-/mnt/cephfs/fio/ben]
    filename=ben
    directory=/mnt/cephfs/fio
    rw=write
    size=104857600

But if I comment out all the histogram parameters, log_hist_msec, log_hist_coarseness, and write_hist_log, then the command succeeds.  

@ndokos I think this is because fio was compiled for pbench-fio without the zlib-devel library being present .  Karl's histogram logging code had a dependency on zlib - he never coded the case where zlib was not present.  This is because compression is important when you have a lot of logs to send back to the client and also zlib package is typically present on every Linux system.  When I look at fio shared libraries, there is no zlib shared library linked into fio.

-=>>rpm -q --whatprovides /usr/local/bin/fio
pbench-fio-2.14-7.x86_64
-=>>ldd /usr/local/bin/fio
	linux-vdso.so.1 =>  (0x00007ffd6a7f3000)
	librt.so.1 => /lib64/librt.so.1 (0x00007ff880852000)
	libaio.so.1 => /lib64/libaio.so.1 (0x00007ff880650000)
	libm.so.6 => /lib64/libm.so.6 (0x00007ff88034d000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007ff880131000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007ff87ff2d000)
	libc.so.6 => /lib64/libc.so.6 (0x00007ff87fb6b000)
	/lib64/ld-linux-x86-64.so.2 (0x00007ff880a65000)

**Comment by ndokos** [comment]
zlib added to pbench-fio spec file as a dependency. Keeping this open until I build an RPM from the official 2.14 tarball. But the current 2.14-9 RPM available from the COPR pbench repo should not have this problem.

**Comment by bengland2** [comment]
Unfortunately this issue has mutated.  We are now having problems after the fio test is over getting the logs transmitted back to the fio --client process, and are not sure why.   I'm on it.    The symptoms of the errors are [in this log file](http://pastebin.com/pa1hLkSB).  I'm trying to isolate the problem by turning off log compression and seeing if it still occurs.

**Comment by bengland2** [comment]
I've reproduced the problem using this job file within 1 minutes, [shown by this pastebin](http://pastebin.com/TUe0i0b4)!  It clearly happens after the test is over and while we are transmitting results back to the test driver from the workload generator fio processes.    I can't turn off log compression because histogram logging didn't implement the code path without zlib, Karl's code prints out an error and exits if you try to do that.  However, if you just turn off histogram logging in the job file and change nothing else, the test runs without error.  Maybe histogram logs have a much bigger record size and fio somehow doesn't deal with this yet.  The first error seems to be "fio: inflate error -3" in  convert_iolog_gz().   I suspect this causes test driver end of socket to lose its place in input stream, resulting in "matching tag" error.

**Comment by portante** [comment]
We want to get a fix into fio itself that pbench relies on to make this scale.

**Comment by bengland2** [comment]
@portante  For the upstream fio description of the problem, [see here](https://github.com/axboe/fio/issues/274).  For the patch that I wrote to work around it, [see here](https://github.com/distributed-system-analysis/pbench/issues/647#issuecomment-331922895).

**Comment by bengland2** [comment]
We have some upstream commits that may fix this issue:

- [ memory corruption bug in log retrieval ](https://github.com/axboe/fio/issues/274#issuecomment-351490691) 
- [allow --eta=never to work](https://github.com/axboe/fio/issues/317#issuecomment-351788459)

have to try it with 500-1000 guests or containers before I'll know.

**Comment by bengland2** [comment]
I think this problem is fixed now, see link [here](https://github.com/axboe/fio/issues/274#issuecomment-351551187)

**Comment by bengland2** [comment]
I consider issue closed.
